### PR TITLE
Unit tests for MiCS

### PR DESCRIPTION
**Pull request by zarzen**
In response to the ask from https://github.com/microsoft/DeepSpeed/pull/2964#issuecomment-1832161865, I added three more unit tests related to MiCS. 

There are two knowledge issues:
- Testing on Torch 2.1.0 triggers `_IllegalWorker` in coalesced all gather. I made changes to ignore this condition. and Currently, I don't know the reason. 
- The MiCS implementation is not working with offloading, so the failure in `TestZeroPartialOffloadConfigSweep` is expected.

### CODE REVIEW

**Comment by tjruwase** [comment]
@zarzen, thanks!

File: tests/unit/runtime/zero/test_zero.py [hunk 1]
```diff
@@ -85,18 +85,25 @@ def test(self, zero_stage):
 
 
 # testing the fix https://github.com/microsoft/DeepSpeed/pull/1227
+@pytest.mark.parametrize("zero_stage", [3, "mics"])
```
Comment (mrwyattii) [inline tests/unit/runtime/zero/test_zero.py:88]:
```suggestion
@pytest.mark.parametrize("mics_enabled", [True, False])
```

File: tests/unit/runtime/zero/test_zero.py [hunk 2]
```diff
@@ -85,18 +85,25 @@ def test(self, zero_stage):
...
 class TestZero3RepeatForwardLoop(DistributedTest):
     world_size = 1
 
-    def test(self, zero_stage=3):
+    def test(self, zero_stage):
```
Comment (mrwyattii) [inline tests/unit/runtime/zero/test_zero.py:92]:
```suggestion
    def test(self, mics_enabled, zero_stage=3):
```

File: tests/unit/runtime/zero/test_zero.py [hunk 3]
```diff
@@ -85,18 +85,25 @@ def test(self, zero_stage):
...
         # force all params to be partitioned by forcing threshold=0
+        mics_shard_size = -1
+        if isinstance(zero_stage, str):
+            zero_stage = 3
+            mics_shard_size = self.world_size
+
```
Comment (mrwyattii) [inline tests/unit/runtime/zero/test_zero.py:98]:
```suggestion
        mics_shard_size = -1
        if mics_enabled:
            mics_shard_size = self.world_size
```

**Review by mrwyattii — CHANGES_REQUESTED** [review]
I think it will be more clear if we parameterize a `mics_enabled` variable rather than use `zero_stage` to switch it on/off. Please update the other test with similar changes. Thanks!

**Comment by zarzen** [comment]
@mrwyattii thanks for suggestions, I updated the implementation accordingly. Besides, as the mics implementation currently is not compatible with offloading, I removed the unittest for `test_zero_offloadpp.py`

**Review by loadams — APPROVED** [review]

**Review by mrwyattii — APPROVED** [review]

**Pull request merged**